Features/rcenon

cenonn opened the pull request:
Added in density and rug overlays for histograms, density plot, and rug plot for 1 variable case. 
Added in 2D histograms and marginal histograms for the 2 variable case.
Redid old plots that were done in seaborn using pure matplotlib where necessary in order to fix compatibility issue. 
Plot can now take a tuple or list for type= 

Zip contains .ipynb for my plotting notebooks.
[plot.zip](https://github.com/dlsun/symbulate/files/1247464/plot.zip)

cenonn commented:
I realized that I have the color always at blue. This will cause problems when plotting multiple histograms for example. Should I be calling color = get_next_color(axes) instead?

dlsun commented on symbulate/results.py:
```diff
@@ -286,22 +287,46 @@ class RVResults(Results):
 
     def plot(self, type=None, alpha=None, normalize=True, jitter=False, 
         bins=30, **kwargs):
+        if type is None:
```
What's the point of this block of code? It doesn't seem to do anything.

cenonn replied:
I have this there since I'm handling the default type situations after going into the different number of variable cases. Should I be handling this differently?

cenonn replied:
I see what you mean here now. Disregard above.

dlsun replied:
Oh I see what you're doing. I think it would be clearer to do:

```
if type is not None:
  if isinstance(type, str):
    ...
  elif isinstance(type, (tuple, list)):
    ...

dlsun commented on symbulate/results.py:
```diff
@@ -286,22 +287,46 @@ class RVResults(Results):
             counts = self._get_counts()
             if type is None:
                 heights = counts.values()
                 if sum([(i > 1) for i in heights]) > .8 * len(heights):
```
The logic in this line (to determine whether to treat the distribution as discrete or continuous) now appears multiple times within this file. You should factor it out as a function and put it in plot.py.

dlsun commented on symbulate/results.py:
```diff
@@ -286,22 +287,46 @@ class RVResults(Results):
+                    type = ("hist",)
+            if alpha is None:
+                alpha = .5
+            if all(x in ("hist",) for x in type):
```
This line is odd. First, why isn't it just `x == "hist"` instead of `x in ("hist",)`. Second, the line suggests that the user would specify something like `type=("hist", "hist", "hist")`. Why would they do that?

dlsun commented on symbulate/results.py:
```diff
@@ -286,22 +287,46 @@ class RVResults(Results):
+                ax = df['X'].plot(kind='hist', normed=normalize, alpha=alpha, bins=bins)
+                df['X'].plot(kind='kde', ax=ax)
+                ax.set_ylabel('')
+            elif all(x in ("hist", "rug") for x in type):
```
I'm not sure I like your approach here. You want the user to be able to superimpose any combination of types on the same graphic. You shouldn't have to anticipate all possible combinations. For example, what if the user wants "hist", "density", and "rug" on the same graphic? You haven't anticipated that combination, so that won't work.

Instead of having these conditionals be mutually exclusive, you should just check if a particular type is in `type`, and then add that type of plot to the `Axis` of the graphic.

dlsun commented on symbulate/results.py:
```diff
@@ -286,22 +287,46 @@ class RVResults(Results):
+                alpha = .5
+            if all(x in ("hist",) for x in type):
+                df = pd.DataFrame({'X':self})
+                ax = df['X'].plot(kind='hist', normed=normalize, alpha=alpha, bins=bins)
```
This use of Pandas seems unnecessary Why not just call `plt.hist(self)` with the appropriate options?

dlsun commented on symbulate/results.py:
```diff
@@ -10,7 +10,6 @@
 import pandas as pd
```
I completely missed this import that you added in a previous commit. I really would like to avoid importing Pandas if possible. It's a large library with a huge overhead. Can you try to do everything in Matplotlib?

dlsun commented on symbulate/results.py:
```diff
@@ -333,56 +358,106 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                ax_joint.scatter(x, y, alpha=alpha)
+                ax_marg_x.hist(x)
+                ax_marg_y.hist(y, orientation='horizontal')
+                plt.setp(ax_marg_x.get_xticklabels(), visible=False)
```
Is this different from `ax_marg_x.xaxis.set_ticklabels([])`?

cenonn replied:
I thought that's how it was supposed to work when I making these plots, but it doesn't work that way. I'm sure why to be completely honest.

dlsun commented on symbulate/results.py:
```diff
@@ -333,56 +358,106 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                res_pivot = pd.pivot_table(res, values='num', index=['Y'],
+                    columns=['X'], aggfunc=np.sum)
+                res_pivot = res_pivot / len(x)
+                res_pivot[np.isnan(res_pivot)] = 0
```
If this is a Pandas DataFrame, you should do `.fillna(0)`. But as discussed elsewhere, you may want to eliminate the use of Pandas altogether.

dlsun commented on symbulate/results.py:
```diff
@@ -333,56 +358,106 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                ax.set_xticklabels(positions)
+            elif all(x in ("violin", ) for x in type) and not discrete_x and discrete_y:
+                fig, ax = plt.subplots(1, 1)
+                res = pd.DataFrame({'X': x, 'Y': y})
```
The use of Pandas for this plot is unnecessary.

dlsun commented on symbulate/results.py:
```diff
@@ -333,56 +358,106 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
-            elif discrete_x == True and discrete_y == False and type == 'violin':
+            elif all(x in ("violin", ) for x in type) and discrete_x and not discrete_y:
                 fig, ax = plt.subplots(1, 1)
                 res = pd.DataFrame({'X': x, 'Y': y})
```
The use of Pandas for this plot is unnecessary.

dlsun commented:
Robert, you should definitely use the color from `get_next_color` rather than just setting the color to  always be blue.

kevindavisross commented:
I haven't looked at the code, but I agree with Dr. Sun that we should reconfigure everything to not use pandas.  From what I can tell, most of the plots should be available in matplotlib, or seaborn.  Please also try different options to generate plots that look nice.

Here are some comments (in no particular order).  I'm still reviewing, but this will get you started.  I have attached a .pdf with some examples.

- Should be no y-axis label or scale when type = "rug" 
- Change default thickness/darkness of 1d "density" curve to be darker
- "density" extends to more extreme x-values than histogram does.  For example, Exponential extends to negative values.  For density, can we restrict the x-axis limits based on the range of actual simulated values? 
- histogram + rug: rug lines don't look great - can we make the marks thicker?  See the [top right plot here](http://seaborn.pydata.org/examples/distplot_options.html)
- density + rug gives an error, but it should be allowed, like the example plot in the previous point.
- For continuous RV: Histogram and "density" y-axis label should be "Density"
- For discrete: Yes, "hist" should be allowed for discrete, e.g. RV(Binomial(1000, 0.5)).sim(10000).plot(type = "hist").  "rug" should be allowed, but would probably not be used (could there be a slight jitter?).
- For discrete: Maybe "density" for discrete should create a plot similar to the plot for a discrete probability model e.g. Binomial(10, .5).plot(), but based on estimated relative frequencies instead of true probabilities.
- 2dhist doesn't look great.  Maybe it's just that the grid lines are there?  See the example [two-dimensional histogram next to `In [12]` here](https://jakevdp.github.io/PythonDataScienceHandbook/04.05-histograms-and-binnings.html)
- 2 scatter plots are not overlaid anymore, but they should be
- marginals option doesn't look great.  Increase number of bins for histogram.  Colors seem very dark in comparison to others.  [See the examples here, in particular the one next to `In [11]`](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html)
- should allow hist2d for two discrete
- for discrete, default marginal should plot impulse, not histogram
- Axis on marginal histogram should be density scale, not count
- For two continuous, there should be a "density2d" type like the [plot next to `In [10]` under "Kernel Density Estimation" here](https://jakevdp.github.io/PythonDataScienceHandbook/04.05-histograms-and-binnings.html)
- For two continuous, when "density2d" is used with "marginal", marginal densities (not histograms) should be plotted, like the [example next to `In [10]` here](https://jakevdp.github.io/PythonDataScienceHandbook/04.14-visualization-with-seaborn.html)
- For one continuous and one discrete, there should be some combination of both "tile" and "hist2d".  Basically, "tile" is "hist2d" with the values not binned.  So for the one continuous, one discrete case, the discrete variable should be unbinned (like in "tile") but the continuous variable should be binned (like in "hist2d").
[Graphics demo.pdf](https://github.com/dlsun/symbulate/files/1257384/Graphics.demo.pdf)

cenonn commented:
Updates:
- Changed style sheet to 'seaborn-colorblind'
- Removed Mosaic
- Updated Marginals to change depending on type of variables and updated styling 
- Added Density Plot for 2 variables
- Removed Pandas from Violin implementation
- Updated type= implementation

Todo:
- Remove Pandas from Tile implementation 
     - Works for most situations, fails with negative values due to multi-index
- Fix Mixed-Tile binning issue

kevindavisross commented:
When I just tried to import Symbulate from this branch, I received the error
```
---> 20 from .plot import configure_axes, get_next_color, discrete_check

ImportError: cannot import name 'discrete_check'

```

So I'm not able to run anything.  Can you fix and recommit?

cenonn commented:
Recommitted. It should work now.

kevindavisross commented:
Here are some quick comments.
- The "density" type for discrete RVs is not plotting the relative frequencies correctly.  They seem to be consistently too large.  The impulse plot is fine.  See the examples in the attached.
- The axes for 2d "density" are not correct.; they don't match up with the ranges of the observed values.  Also, when "marginal" is added, there seems to be a mismatch in the axes of the marginal plots and the joint plot.  See the examples in the attached.
- Calling "tile" for two continuous returns an error.  Maybe it should return "scatter" instead plus a warning like "type="tile" is only appropriate for discrete variables".
- For "tile", have the tick marks correspond to the center of the rectangles, instead of the endpoints.  (It was this way originally; maybe switching from pandas changed it.)
- Syntax questions for @dlsun : we are using "hist" for both 1d and 2d histograms.  For consistency, should we replace "tile" with "impulse" for 2d discrete?
- Add a colorbar (like for "tile") to 2d "hist" and 2d "density"
- Can we move the colorbar to the left of the plot when "marginal" is added?
- We've lost all the axis labels.  For "impulse", y-axis label should be "Relative frequency".  For "hist" and "density" y-axis label should be "Density".  For 2d plots, if you can put a title on the colorbar, it should say "Relative frequency" for "tile" or "Density" for "hist"/"density".
- I personally prefer the white background without gridlines, but maybe @dlsun feels otherwise?
- The alpha parameter doesn't seem to work as well as before.  There doesn't seem to be much difference between alpha = 0.1, 0.2, 0.3, and after 0.5 or so there doesn't seem to be much transparency at all.  Could just be my eyes.

@cenonn , I'm emailing you a Jupyter notebook with lots of examples you can use to test graphics.  Some of the examples are in the attached.
[Graphics demo2.pdf](https://github.com/dlsun/symbulate/files/1278828/Graphics.demo2.pdf)

cenonn commented:
I've fixed the colorbars, the 2d density, and labels so far. 

I'm still waiting for Dr. Sun's assistance on fixing the tile issues so I haven't touched it since. I've looked into trying to fix the density for discrete 1 RV and I'm pretty sure it's something to do with how scipy is estimating the density. I'm gonna try and come up with a fix.

kevindavisross commented:
Quick comment: for discrete RV the "density" plot is basically the same as the impulse plot, just with connected dots instead of spikes.  Nothing should be estimated differently, it's the same data in both plots.  Look at the code for impulse to find the data. Or just pull the values from .tabulate(normalize=True).

kevindavisross commented:
Density plot for discrete RV seems to be fixed.  Thanks!

Tile plot has some issues, but it's more than just the axes.  See the attached example.  The colors of the cells don't correspond to the simulated relative frequencies.  I really like the tile plot as an alternative to scatter+jitter for 2 discrete, so it would be great if we could get this working properly.

A few other small changes
- I think the colorbar for 2d density is based on counts; rescale that it's on a density scale (so that [density value * area of rectangle] represents probability) 
- y-axis label for "density" type for discrete RV should be "Relative frequency" (same as for "impulse")
- y-axis label for "density" type for continuous RV should be "Density" (same as for "hist")
[Tile demo.pdf](https://github.com/dlsun/symbulate/files/1283478/Tile.demo.pdf)

dlsun commented:
What help do you need with the tiles?

By the way, I can't seem to reproduce the speed issues that you were having. Even when I'm on my branch, the simulations seem to run in a reasonable amount of time: 700 ms for 10000 simulations of a Normal(0, 1) random variable.

cenonn commented:
You recommended that I don't use a multi-index or Pandas to format the data
into a way that matplotlib can use it. I'm not sure how to do so
otherwise. You may have also mentioned a solution for binning the data
better for the mixed-tile plot, but I don't recall what you said.

The speed issue seemed to go away when I was using my desktop later that
evening. My laptop may have been running slow that morning for some reason.

On Thu, Sep 7, 2017 at 11:55 AM dlsun <notifications@github.com> wrote:

> What help do you need with the tiles?
>
> By the way, I can't seem to reproduce the speed issues that you were
> having. Even when I'm on my branch, the simulations seem to run in a
> reasonable amount of time: 700 ms for 10000 simulations of a Normal(0, 1)
> random variable.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dlsun/symbulate/pull/46#issuecomment-327891898>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/APuKkw2mqaBeuurodayRc5coywz7ZkVQks5sgDwsgaJpZM4PAyFw>
> .
>

kevindavisross commented:
Whatever code is in place now for tile isn't working properly.  See my previous comment with an example of colors not matching up with frequencies.

I've also attached a copy of a tile plot from old code that seemed to work fine (except I had wanted the axes to be number lines instead of just categories).  When I try to run this same example now, I get the error in the attached.
[Old tile plot with new error.pdf](https://github.com/dlsun/symbulate/files/1285874/Old.tile.plot.with.new.error.pdf)
[Old tile plot.pdf](https://github.com/dlsun/symbulate/files/1285875/Old.tile.plot.pdf)

kevindavisross commented:
Tile looks much better and seems to be working now.  Thanks!

Please continue to see what you can do re tile/hist combo for discrete X and continuous Y.

A few small changes
- The colorbar for 2d hist is still on a count scale; please rescale to density.  (The colorbar for 2d density does have the correct density scale.) 
- y-axis label for ["density", "rug"] should be "Density".  (But make sure "rug" by itself continues to have no y-axis label or scale.)
- Calling "tile" when not two discrete variables returns an error. Maybe it should return "scatter" instead plus a warning like "type="tile" is only appropriate for two discrete variables".

cenonn commented:
Mixed-tile is supposed to be a mix between a 2d histogram and tile plot. It's supposed to handle the situation where we have 1 continuous and 1 discrete variable and we want to plot it in a way that will look like a tile plot. To do this, I'm binning the 1 continuous variable and then trying to put the counts into a matrix to pass into either imshow or matshow. The problem is that I can't figure out how to put the counts into the matrix like how I did it with tile.

cenonn commented:
@kevindavisross  I've also fixed the issues you've mentioned.

dlsun commented on symbulate/results.py:
```diff
@@ -330,61 +364,205 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                counts = self._get_counts()
+                x_uniq = np.unique(x)
+                y_uniq = np.unique(y)
+                xmin = min(x)
```
Please write this as `xmin, xmax = min(x), max(x)` and ditto for y.

dlsun commented on symbulate/results.py:
```diff
@@ -330,61 +364,205 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+
+                intensity = np.zeros(shape=(len(y_pos), len(x_pos)))
+                for key, val in counts.items():
+                    intensity[key[1] - ymin][key[0] - xmin] = val / nums
```
This code won't work if x and y contain non-integer values. You should really use the values of x_pos and y_pos, since those are guaranteed to be 0 to (xmax - xmin).

cenonn replied:
In the 2 discrete variable case, won't all the values be integers though? If 'tile' is passed into plot when the variables are continuous, there is a check that will change the type to scatter.

dlsun replied:
Not necessarily. A discrete random variable could take on the values 0, 0.5, 1.0, 1.5, ...

dlsun commented on symbulate/results.py:
```diff
@@ -330,61 +364,205 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                            i += 1
+                        x_new.append(x_bin[i])
+                    x = x_new
+                    counts = count_var(list(zip(x, y)))
```
I don't understand why you don't use np.digitize here.

dlsun commented on symbulate/results.py:
```diff
@@ -495,50 +504,50 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                ymax, ymin = max(y), min(y)
+                x_pos = list(range(len(x_uniq)))
+                y_pos = list(range(len(y_uniq)))
+                x_map = dict(zip(x_uniq, x_pos))
```
I think this is wrong. This seems to assume that x_uniq is sorted?

dlsun replied:
Nvm. `np.unique()` returns the sorted unique values. This is worth a comment above.

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +40,7 @@ def plot(*args, **kwargs):
     except:
         plt.plot(*args, **kwargs)
         
+def discrete_check(heights):
```
Rename function to `is_discrete`.

dlsun commented on symbulate/results.py:
```diff
@@ -285,23 +288,54 @@ def truncate(result):
+                else:
+                    density = gaussian_kde(self)
+                    density.covariance_factor = lambda: 0.25
+                    density._compute_covariance()
```
The logic in this block of code appears multiple times in this file (again for 2 variables). Try factoring out into plot.py?

dlsun commented on symbulate/results.py:
```diff
@@ -330,61 +365,216 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                cbar = plt.colorbar(mappable=den, cax=caxes)
+                caxes.yaxis.set_ticks_position('left')
+                cbar.set_label('Density')
+                caxes.yaxis.set_label_position("left")
```
Please use quotes consistently. You called it `'left'` above, but `"left"` here.

dlsun commented on symbulate/results.py:
```diff
@@ -330,61 +365,216 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                y_pos = range(len(y_uniq))
+                x_map = dict(zip(x_uniq, x_pos))
+                y_map = dict(zip(y_uniq, y_pos))
+                x = np.vectorize(x_map.get)(x)
```
This is not good. It looks like this line of code overrides the definition of `x`, which is very dangerous. This has to be given a different name.

dlsun commented on symbulate/results.py:
```diff
@@ -330,61 +365,216 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                caxes.yaxis.set_ticks_position('left')
+                cbar.set_label('Relative Frequency')
+                caxes.yaxis.set_label_position("left")
+            elif 'mixed-tile' in type:
```
The code in this block is repetitive: (1) you write essentially the same code for x and y, and (2) a lot of the logic is exactly the same as for "tile" plots.

As soon as you feel like you are writing the exact same code more than once, you should think about factoring it out. In this case, I think almost all of the logic for tile and mixed-tile plots should be factored into plot.py.

dlsun commented on symbulate/results.py:
```diff
@@ -330,61 +365,216 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                den = ax.imshow(Z.reshape(Xgrid.shape), origin='lower', cmap='Blues',
+                          aspect='auto', extent=[xmin, xmax, ymin, ymax]
+                )
+                if 'marginal' not in type:
```
Hmm...I feel like I've seen this code before. Can you factor it out?

dlsun commented on symbulate/results.py:
```diff
@@ -330,61 +365,216 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                    bins = 30
+
+            if 'tile' in type and (not discrete_x or not discrete_y):
+                print('type=\'tile\' is only valid for 2 discrete variables')
```
If you used double quotes " ", you wouldn't need to escape the single quotes ' ' inside the string.

dlsun commented on symbulate/results.py:
```diff
@@ -330,61 +365,216 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                    violins = ax.violinplot(dataset=values, showmedians=True)
+                    ax.set_xticks(np.array(positions) + 1)
+                    ax.set_xticklabels(positions)
+                else: 
```
Do you see how the code in this `else` block is basically the same as the above code, except with "x" replaced by "y"? This is a good candidate for code to factor into plot.py. You can write a function `make_violin_plot()` that takes in the data and which axis ("x" or "y") is discrete and then generates the appropriate plot.

I encourage you to also write a helper function `set_ticks()` in plot.py that takes in the positions, labels, and which axis ("x" or "y").

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,117 @@ def plot(*args, **kwargs):
+        v_lab = [min(v)]
+        for i in range(len(v_bin) - 1):
+            v_lab.append(0.5 * (v_bin[i] + v_bin[i+1]))
+        v_pos = range(len(v_lab))
```
Suggestion: Change `v_pos` to be [0.5, 1.5, 2.5, ...] instead of [0, 1, 2, ...] and change `v_lab` to be the actual endpoints of the bins (instead of the average.)

cenonn replied:
Changing v_lab to the actual ends points makes it the same array as v_bin, doesn't it?

dlsun replied:
Yes, I believe that's right.

dlsun commented on symbulate/results.py:
```diff
@@ -313,78 +337,103 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                cbar, caxes = add_colorbar(fig, type, den)
+            elif 'tile' in type:
+                hm, x_lab, y_lab, x_pos, y_pos = make_tile(x, y, bins, discrete_x, discrete_y, ax)
+                setup_ticks(x_pos, x_lab, ax, 'x')
```
Why not just call `setup_ticks` instead `make_tile`? Then `make_tile` wouldn't have to return all of these objects.

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,117 @@ def plot(*args, **kwargs):
+        ax.set_xticks(np.array(positions) + 1)
+        ax.set_xticklabels(positions)
+    elif axis == 'y':
+        for i in positions:
```
The code for `axis == 'y'` is almost exactly the same as the code for `axis == 'x'`. Can you factor this out into a function?

(Also, note that you can use your `setup_ticks` function above.)

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,117 @@ def plot(*args, **kwargs):
+        
+    cbar = plt.colorbar(mappable=mappable, cax=caxes)
+    caxes.yaxis.set_ticks_position('left')
+    cbar.set_label('Density')
```
Is it always guaranteed to be a density (and not, say, a count)?

cenonn commented:
[plotting.zip](https://github.com/dlsun/symbulate/files/1291068/plotting.zip)

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,117 @@ def plot(*args, **kwargs):
+        vp.set_linewidth(1)
+    return violins
+
+def get_violin(data, positions, ax, axis):
```
This function is not really necessary, since it only gets called once.

What I was hoping you'd do is factor out the code under `if axis == 'x'` and `elif axis == 'y'` into a separate function. Look how similar they are; the only difference is that the roles of 0 and 1 are swapped. Can you please move this code back into `make_violin`?

What do you think about changing the next 7 lines to just 2 lines?
```
i, j = (0, 1) if axis == 'x' else (1, 0)
values = [data[data[:, i] == pos, j].tolist() for pos in positions]
```
Feel free to push back on this suggestion if you don't think this code is actually better / clearer.

cenonn replied:
This makes sense. It wasn't clear to me how to handle the conditional while I was working on this earlier.

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,117 @@ def plot(*args, **kwargs):
+    elif axis == 'y':
+        ax_marg.hlines(key, 0, val, color=color, alpha=alpha)
+
+def density2D(x, y, ax):
```
Should this be called `make_density2D` to be consistent with the other function names?

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,117 @@ def plot(*args, **kwargs):
+    setup_ticks(y_pos, y_lab, ax, 'y')
+    return hm
+
+def setup_ticks(pos, lab, ax, axis):
```
Please move `setup_ticks` either above or below all of the different `make_*` functions.

dlsun commented on symbulate/results.py:
```diff
@@ -313,78 +337,100 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                    bins = 10
+                else:
+                    bins = 30
+            if 'tile' in type and (not discrete_x and not discrete_y):
```
Wouldn't it make more sense to move this check inside of the `elif 'tile' in type:` block below? 

(I understand why you're doing it -- it's to change the type to scatter for the user. But I would almost prefer that you just raise an Exception and suggest that the user try again with `type='scatter'`.)

dlsun commented on symbulate/results.py:
```diff
@@ -313,78 +337,100 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
             if alpha is None:
                 alpha = .5
+            if bins is None:
+                if 'tile' in type:
```
One liner: `bins = 10 if 'tile' in type else 30`

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,117 @@ def plot(*args, **kwargs):
+	return density
+    
+def add_colorbar(fig, type, mappable, label):
+    if 'marginal' not in type:
```
It looks like this is only getting called if there's a colorbar. Is this the desired behavior?

cenonn replied:
Yeah that's the intent here. We want the colorbar to on the left of the plot and the ideal way to do this is to give it its own axis. The reason for the if statement here is to adjust the size of the colorbar when there are marginal plots.

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,117 @@ def plot(*args, **kwargs):
+
+def make_violin(data, positions, ax, axis, alpha):
+    violins = get_violin(data, positions, ax, axis)
+    for part in violins['bodies']:
```
I think the usage of the variable name `part` here is confusing, as it seems to mean something different from `part` below. According to the usage below, `'bodies'` would be the `part`. So I'd suggest that you call this `body_part` or `bp`? I'm not sure; use your best judgement here.

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,117 @@ def plot(*args, **kwargs):
+    return violins
+
+
+def marginal_impulse(count, height, color, ax_marg, alpha, axis):
```
`height` seems unnecessary here. Isn't `height` just `count.values()`?

In this case, I'd prefer 
```
key, val = count.keys(), count.values()
...
```

Also, is there any compelling reason that they need to be cast to lists?

cenonn replied:
I convert to list to get around a type error. (TypeError: float() argument must be a string or a number, not 'dict_keys')

dlsun commented on symbulate/results.py:
```diff
@@ -313,78 +337,96 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                cbar, caxes = add_colorbar(fig, type, den, 'Density')
+            elif 'tile' in type:
+                if not discrete_x and not discrete_y:
+                   raise Exception("type='tile' is only valid for 2 discrete variables" )
```
That's not true, is it? It seems like your tile plot will work if there's 1 discrete and 1 continuous, or even if both variables are continuous. If so, you may just want to abandon this error message altogether.

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,108 @@ def plot(*args, **kwargs):
+        ax.set_yticklabels(lab)
+    
+def add_colorbar(fig, type, mappable, label):
+    if 'marginal' not in type:
```
Please add a comment here explaining the purpose of this, like what you explained to me.

dlsun commented on symbulate/results.py:
```diff
@@ -313,78 +337,96 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                caxes.set_yticklabels(new_labels)
+            elif 'density' in type:
+                den = make_density2D(x, y, ax)
+                cbar, caxes = add_colorbar(fig, type, den, 'Density')
```
If you don't use `cbar` and `caxes` again, don't save them. (Just write `add_colorbar()`.)

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,108 @@ def plot(*args, **kwargs):
+    caxes.yaxis.set_ticks_position('left')
+    cbar.set_label(label)
+    caxes.yaxis.set_label_position('left')
+    return cbar, caxes
```
Why do you return `cbar`? It doesn't seem to get used.

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,108 @@ def plot(*args, **kwargs):
+    return counts
+    
+def compute_density(values):
+	density = gaussian_kde(values)
```
The indent here doesn't match the rest of the file.

dlsun commented on symbulate/results.py:
```diff
@@ -313,78 +337,96 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                res = np.array(self)
+                if discrete_x and not discrete_y:
+                    positions = sorted(list(x_count.keys()))
+                    violins = make_violin(res, positions, ax, 'x', alpha)
```
If you don't use `violins`, don't store it.

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,108 @@ def plot(*args, **kwargs):
+	density._compute_covariance()
+	return density
+
+def setup_ticks(pos, lab, ax, axis):
```
Suggestion: I believe that you can set ticks alternatively using `ax.xaxis.set_ticks` and `ax.yaxis.set_ticks`. Then you wouldn't have to pass in both `ax` and `axis`. You can just pass around either `ax.xaxis` or `ax.yaxis`, which would specify which axis you want to modify, and get rid of the `axis=` argument. Can you explore this to see if it would work?

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,108 @@ def plot(*args, **kwargs):
+        v_pos = range(len(v_lab))
+        v_vect = np.digitize(v, v_lab)
+    else:
+        v_lab = np.unique(v)
```
Comment that this is sorted.

dlsun commented on symbulate/plot.py:
```diff
@@ -40,4 +41,108 @@ def plot(*args, **kwargs):
+    if not discrete:
+        v_lab = np.linspace(min(v), max(v) + 1, bins)
+        v_pos = range(len(v_lab))
+        v_vect = np.digitize(v, v_lab)
```
I'm pretty sure that the plots that this code will produce are off by 1. Can you please double check this?

The lowest number when you call `np.digitize` is going to be 1 (because every value in `v` is >= min(v)). But `v_pos` starts at 0. Am I missing something here?

cenonn replied:
For clarification, do you mean that the labels are shifted over by 1 or that the actual positions within the intensity matrix will be shifted over by 1?

dlsun replied:
It doesn't matter if the labels and intensity matrix are **both** shifted over by 1. The problem is that they're inconsistent with each other.

As @kevindavisross pointed out, the labels are still at the midpoints of the tiles. They should be at the endpoints of the tiles, and you should double check that the numbers are correct (i.e., not shifted over by 1) by plotting a few examples where you know what the plot should look like.

kevindavisross commented:
The tick marks on the continuous axis are still at midpoints, not endpoints.

cenonn commented:
From what I can tell, It seems to be working. Based on looking at the plots and a bunch of print statements, I don't **think** the labels and matrix are inconsistent with each other.

dlsun commented:
Robert, it still looks to me like you're labeling the midpoints of the tiles, not the endpoints of the tiles. Unfortunately, I'm at work and don't have any ability to test it right now. Are you sure you have this right?

dlsun commented:
To be clear, you do want to label the midpoints in the case the variable is discrete. But in the case that the variable is continuous, it would be better to label the endpoints.

cenonn commented:
That's how it is currently behaving unless I'm understanding the issue incorrectly.

dlsun commented:
@kevindavisross Can you please take a look and see if this is working as you intended?

kevindavisross commented:
See the attached examples.  On the surface, the first tile plot looks right, except that the y-axis ranges from 0 to 3 when it only should range from 0 to 2.  So it doesn't look the greatest with the unnecessary white space.

But this causes a bigger problem when adding the marginal distribution; on the marginal plot it appears that the values of Y range from 0 to 3, when should only be 0 to 2.  

A non-tile issue:, when "marginal" is called by itself, the axes are just [0, 1], not the range of the values of the variable.  (I admit I didn't test this particular feature too closely.)
 
[mixed tile.pdf](https://github.com/dlsun/symbulate/files/1293766/mixed.tile.pdf)

dlsun commented on symbulate/plot.py:
```diff
@@ -87,19 +87,30 @@ def setup_tile(v, bins, discrete):
         v_vect = np.vectorize(v_map.get)(v)
     return v_vect, v_lab, v_pos
 
+def adjust_tile(intensity, lab, pos, bins, axis):
```
It doesn't seem like this function should be necessary. Why not just get `v_lab` and `v_pos` correct in the function `setup_tile`?

kevindavisross commented:
I agree.  I'm merging now.

The issue with "marginal" affects a few of plot types.  Maybe we just shouldn't advertise "marginal" until we get in cleaned up.

Thanks, Robert, for all your hard work!
